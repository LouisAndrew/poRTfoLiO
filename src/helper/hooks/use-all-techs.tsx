import { useEffect, useState } from 'react'
import { useStaticQuery, graphql } from 'gatsby'
import { FixedObject } from 'gatsby-image'

import { Tech } from 'helper/schema/Tech'

const useAllTechs = () => {
    const data = useStaticQuery(graphql`
        query {
            techData: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_name
                        }
                    }
                }
            }
            imgXS: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(width: 40, height: 40, quality: 100) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
            imgS: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(width: 60, height: 60, quality: 100) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
            imgM: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(
                                        width: 100
                                        height: 100
                                        quality: 100
                                    ) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
            imgL: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(
                                        width: 120
                                        height: 120
                                        quality: 100
                                    ) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    `)

    const [techs, setTechs] = useState<Tech[]>([])

    useEffect(() => {
        const { techData, imgS, imgM, imgL, imgXS } = data as any

        console.log({ techData, imgS, imgM, imgL, imgXS })

        const temp: Tech[] = techData.edges.map((edge: any, i: number) => {
            const techName = edge.node.frontmatter.tech_name
            const xs =
                imgXS.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const s =
                imgS.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const m =
                imgM.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const l =
                imgL.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const xl: FixedObject[] = []

            return {
                techName,
                techLogo: {
                    xs,
                    s,
                    m,
                    l,
                    xl,
                },
            }
        })

        setTechs(temp)
    }, [])

    /**
     * Get a single tech by filtering its name
     * @param name Name of the technology
     * @returns the queried tech, and null if the tech is not available..
     */
    const getTechByName = (name: string) => {
        const techArr = techs.filter(
            tech => tech.techName.toLowerCase() === name.toLowerCase()
        )

        if (techArr.length === 0) {
            return null
        } else {
            return techArr[0]
        }
    }

    return { getTechByName, techs }
}

export { useAllTechs }
