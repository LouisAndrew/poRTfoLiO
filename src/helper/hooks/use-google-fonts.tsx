import { useEffect } from 'react'

/**
 * Custom hook to embed google font script / api to html head
 */
const useGoogleFonts = () => {
    const LINK_URL =
        'https://fonts.googleapis.com/css2?family=Karla:wght@400;700;800&family=Nunito:wght@300;400;700;800&display=swap'

    const LINK_PRECONNECT_URL = 'https://fonts.gstatic.com'

    /**
     * Function to create link elements
     * @returns array of link element
     */
    const createLinks = () => {
        const fontLink = document.createElement('link')
        const preconnectLink = document.createElement('link')

        fontLink.href = LINK_URL
        fontLink.rel = 'stylesheet'

        preconnectLink.href = LINK_PRECONNECT_URL
        preconnectLink.rel = 'preconnect'

        return [fontLink, preconnectLink]
    }

    useEffect(() => {
        const links = createLinks()

        // Append links to document's head tag
        links.forEach(link => {
            document.head.appendChild(link)
        })

        return () => {
            // cleanup, remove links from document's head tag
            links.forEach(link => {
                document.head.removeChild(link)
            })
        }
    }, [])
}

export { useGoogleFonts }
