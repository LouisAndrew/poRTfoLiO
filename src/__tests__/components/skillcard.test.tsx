import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import { render, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom'

import Skillcard from 'components/skillcard'

// import * as Gatsby from 'gatsby'
import * as CustomHook from 'helper/hooks/use-all-techs'

// const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery')
const useAllTechs = jest.spyOn(CustomHook, 'useAllTechs')
const mockTechName = 'React'
const mockTech = {
    techName: mockTechName,
    techLogo: {
        s: [],
        xs: [],
        m: [],
        l: [],
        xl: [],
    },
}

describe('Skillcard', () => {
    const Element = <Skillcard techName={mockTechName} />
    const ElementNotAvailable = <Skillcard techName="Java" />

    afterEach(cleanup)
    beforeEach(() => {
        useAllTechs.mockImplementationOnce(() => ({
            techs: [mockTech],
            getTechByName: name => {
                if (name === mockTechName) {
                    return mockTech
                } else {
                    return null
                }
            },
        }))
        // useStaticQuery.mockImplementationOnce(() => ({
        //     techName: {
        //         edges: [
        //             {
        //                 node: {
        //                     frontmatter: {
        //                         tech_name: mockTechName,
        //                     },
        //                 },
        //             },
        //         ],
        //     },
        //     imgXS: {
        //         edges: [
        //             {
        //                 node: {
        //                     frontmatter: {
        //                         tech_logo: {
        //                             childImageSharp: {
        //                                 fixed: {},
        //                             },
        //                         },
        //                     },
        //                 },
        //             },
        //         ],
        //     },
        //     imgS: {
        //         edges: [
        //             {
        //                 node: {
        //                     frontmatter: {
        //                         tech_logo: {
        //                             childImageSharp: {
        //                                 fixed: {},
        //                             },
        //                         },
        //                     },
        //                 },
        //             },
        //         ],
        //     },
        //     imgM: {
        //         edges: [
        //             {
        //                 node: {
        //                     frontmatter: {
        //                         tech_logo: {
        //                             childImageSharp: {
        //                                 fixed: {},
        //                             },
        //                         },
        //                     },
        //                 },
        //             },
        //         ],
        //     },
        //     imgL: {
        //         edges: [
        //             {
        //                 node: {
        //                     frontmatter: {
        //                         tech_logo: {
        //                             childImageSharp: {
        //                                 fixed: {},
        //                             },
        //                         },
        //                     },
        //                 },
        //             },
        //         ],
        //     },
        // }))
    })

    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(Element, div)
    })

    it('Should render the skillcard given the correct tech name', () => {
        const { queryByTestId, queryByText } = render(Element)

        const wrapper = queryByTestId('wrapper')
        const text = queryByText(mockTechName)

        expect(wrapper).toBeInTheDocument()
        expect(text).toBeInTheDocument()
    })

    it('Should not render the skillcard when the given tech name is not available in the graphql store', () => {
        const { queryByTestId, queryByText } = render(ElementNotAvailable)

        const wrapper = queryByTestId('wrapper')
        const text = queryByText(mockTechName)

        expect(wrapper).not.toBeInTheDocument()
        expect(text).not.toBeInTheDocument()
    })

    it('matches snapshot', () => {
        const run = true

        expect(run).toBeTruthy()

        if (run) {
            const tree = renderer.create(Element).toJSON()
            expect(tree).toMatchSnapshot()
        }
    })
})
