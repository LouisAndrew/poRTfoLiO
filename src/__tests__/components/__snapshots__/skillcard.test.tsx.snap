// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skillcard matches snapshot 1`] = `
<div
  className="skillcard css-1dp33er"
  data-testid="wrapper"
>
  <div
    className="css-12os43k"
  >
    <div
      className="css-1x5vnaj"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": undefined,
            "overflow": "hidden",
            "position": "relative",
            "width": undefined,
          }
        }
      >
        <picture>
          <source
            media="(max-width: 640px)"
          />
          <source
            media="(max-width: 832px) and (orientation: landscape)"
          />
          <source
            media="(max-width: 1040px)"
          />
          <source
            media="(min-width: 1952px)"
          />
          <source />
          <img
            alt=""
            loading="lazy"
            onLoad={[Function]}
            style={
              Object {
                "height": "auto",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source media=\\"(max-width: 640px)\\" srcset=\\"undefined\\" /><source media=\\"(max-width: 832px) and (orientation: landscape)\\" srcset=\\"undefined\\" /><source media=\\"(max-width: 1040px)\\" srcset=\\"undefined\\" /><source media=\\"(min-width: 1952px)\\" srcset=\\"undefined\\" /><source srcset=\\"undefined\\" /><img loading=\\"lazy\\" src=\\"\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="css-m4qzk6"
  >
    React
  </div>
</div>
`;
